### DISCUSSION TITLE
Infrastructure requirements for a customer PoC request

### DISCUSSION THREAD

**Discussion by djdkochasoft** [frePPLe/frepple#514] [Q&A]
Hey there,

We are a software vendor contracted to provide a supply planning solution for our client.
Hence we are evaluating Frepple as a potential solver back end for our supply planning application.
We have been modeling Frepple's capabilities for the client's specific use cases, in a PoC fashion.

We have built a model which matches the client's master data models with the following data counts:
- Items (UPC) : 27,000
- Locations (vendors) : 160
- Customers (DCs) : 5
- Operations : 200,000+ (1 routing + 2 fixed time operations for each item)
- Resources : 160

Before presenting this to our client we want to conduct a performance test. In our 1st performance test using 1 demand, it takes around 10 mins to plan. Our objective is to do an evaluation of the solver performance against 50k demands followed by 1 million demands and evaluate if the runtime scales linearly.

Below is the log file from the run with 1 demand.
[Log file 8 - frePPLe.txt](https://github.com/frePPLe/frepple/files/8211888/Log.file.8.-.frePPLe.txt)

Here are our open questions:
1. Please advise us on how to size our infrastructure for our PoC (e.g. are there baseline requirements and how they should scale for 1 million demands)
2. Is there a way to split our data model to use parallel processing to get a faster output? (e.g. using low code to identify discreet supply chains within the model)

**Comment by jdetaeye** [comment]
10 minutes for 3 demands is way too long. You need to run with loglevel 2 to see why that goes so inefficiently.  Only after that can you perform a real performance test.
The rest of the log file looks pretty normal.

> Please advise us on how to size our infrastructure for our PoC (e.g. are there baseline requirements and how they should scale for 1 million demands)

There is no generic advice here.  Models vary widely in size and complexity.

> Is there a way to split our data model to use parallel processing to get a faster output? (e.g. using low code to identify discreet supply chains within the model)

The planning engine will automatically parallelize the planning where possible.   In the very majority of situations that is sufficient.  

If the split the planning problem in a certain logical dimension (eg different business units, different unrelated product lines, etc...) you can use different scenarios to divide the data & process & calculations.

**Reply by djdkochasoft** [reply]
Thanks for your response.
After expanding the log it would seem a majority of the time is spent on "safety stock replenishment passes" (starts from cluster 10475)
We haven't configured anything related to safety stock. Is there a way to avoid the time spent on this?

**Reply by jdetaeye** [reply]
The question is not "how to avoid" but "why does it take so long"...     Going through the 73k buffers and scanning whether they need a safety stock replenishment should not take that long.          There must be something incorrect or inefficiently modeled in the supply path.

As a quick test, you can try running an unconstrained plan. That should go through very fast and the resulting plan will give you an indication of the bottleneck constraints.    If also the unconstrained plan takes very long, you definitely have something bad in your supply paths,

**Reply by djdkochasoft** [reply]
Hello!
We managed to get a very good runtime for 20k records. Thanks for your support.

The other main feature we would need to evaluate is running multiple scenarios parallely.
We see that the feature is available as seen in this screenshot ![Capture](https://user-images.githubusercontent.com/75317758/158069758-d0db3a36-4af5-4a63-9150-9bd13639d1c4.PNG)

However, creating a copy of the default database results in a "Database copy failed" message. 

1. Do you have any suggestions on how to resolve this error?
2. The help text says "Use the dropdown in the upper right corner of the screen to select which scenario you're working in.".. Such an option is not available on any of the screens. How do we switch between scenarios?

**Reply by jdetaeye** [reply]
>> Do you have any suggestions on how to resolve this error?

Check out the log file /var/log/apache2/error.log for clues.   If nothing suspicious is found, try running on the command line "frepplectl scenario_copy default scenario1. It may give a better error message.

Very likely there is some database configuration or permission causing the trouble. 
Another gotcha is that the version of the postgresql client isn't matching the database server postgresql version (these must match for the pg_dump and pg_restore utilities to work correctly)

>>The help text says "Use the dropdown in the upper right corner of the screen to select which scenario you're working in.".. Such an option is not available on any of the screens. How do we switch between scenarios?

It'll appear after a successful scenario copy.
